OBSDOCS-668: monitoring 4.15 RN new features (#71003)
openshift/openshift-docs: release_notes/ocp-4-15-release-notes.adoc

[id="ocp-4-15-monitoring"]
=== Monitoring

The in-cluster monitoring stack for this release includes the following new and modified features.

[id="ocp-4-15-monitoring-updates-to-monitoring-stack-components-and-dependencies"]
==== Updates to monitoring stack components and dependencies
This release includes the following version updates for in-cluster monitoring stack components and dependencies:

* Alertmanager to 0.26.0
* kube-state-metrics to 2.10.1
* node-exporter to 1.7.0
* Prometheus to 2.48.0
* Prometheus Adapter to 0.11.2
* Prometheus Operator to 0.70.0
* Thanos Querier to 0.32.5

[id="ocp-4-15-monitoring-changes-to-alerting-rules"]
==== Changes to alerting rules

[NOTE]
====
Red{nbsp}Hat does not guarantee backward compatibility for recording rules or alerting rules.
====

* *Changed*

> ** Modified the `NodeClockNotSynchronising` alerting rule to also consider the PTP protocol.

Reviewer: ```suggestion
** Modified the `NodeClockNotSynchronising` and `NodeClockSkewDetected` alerting rule to also consider the PTP protocol.
```

Reviewer 2: Maybe?
```suggestion
** Disabled the `NodeClockNotSynchronising` and `NodeClockSkewDetected` alerting rules when the Precision Time Protocol (PTP) is in use.
```

Author: Changed per @simonpasquier's suggestion

[id="ocp-4-15-monitoring-new-metrics-server-to-access-metrics-API-technology-preview"]
==== New Metrics Server component to access the Metrics API (Technology Preview)
This release introduces a Technology Preview option to add a Metrics Server component to the in-cluster monitoring stack. 

> If enabled, Metrics Server, rather than Prometheus Adapter, collects resource metrics and exposes them in the `metrics.k8s.io` Metrics API service for use by other tools and APIs.

Reviewer 2: It might be unclear if you need to enable the metrics server explicitly or not. In practice, metrics server is automatically installed instead of Prometheus adapter when the FeatureGate CR is configured with `TechPreviewNoUpgrade`.

https://docs.openshift.com/container-platform/4.14/nodes/clusters/nodes-cluster-enabling-features.html

Author: Added verbiage about how Metrics Server is installed.

For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#metricsserverconfig[MetricsServerConfig] in the config map API reference for the Cluster Monitoring Operator.

[id="ocp-4-15-monitoring-new-feature-to-send-exemplar-data-to-remote-write-storage-for-user-defined-projects"]
==== New feature to send exemplar data to remote write storage for user-defined projects
With this release, your user-defined projects can now use remote write to send exemplar data scraped by Prometheus to remote storage.
To use this feature, configure remote write using the `sendExemplars` option in the `RemoteWriteSpec` resource.
For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#remotewritespec[RemoteWriteSpec] in the config map API reference for the Cluster Monitoring Operator.

[id="ocp-4-15-monitoring-impoved-alert-querying-for-user-defined-projects"]
==== Improved alert querying for user-defined projects
With this release, applications in your user-defined projects now have API access to query alerts for application namespaces via the rules tenancy port for Thanos Querier. 

> Specifically, you can now constuct queries that access the `/api/v1/alerts` endpoint via port 9093 for Thanos Querier, provided that the HTTP request contains a `namespace` parameter.

Reviewer: ```suggestion
Specifically, you can now construct queries that access the `/api/v1/alerts` endpoint via port 9093 for Thanos Querier, provided that the HTTP request contains a `namespace` parameter.
```

Author: Fixed

In previous releases, the rules tenancy port for Thanos Querier did not provide API access to the `/api/v1/alerts` endpoint.

[id="ocp-4-15-monitoring-prometheus-updated-to-tolerate-scrape-time-jitters"]
==== Prometheus updated to tolerate scrape time jitters to reduce TSDB storage requirements
With this release, the default Prometheus configuration in the in-cluster monitoring stack has been changed so that jitters are tolerated at scrape time, which improves chunk compression for data storage.
As a result, the time series database uses significantly less disk space.

> Note that this change can reduce the accuracy of derived data, such as the running average of a time series.

Reviewer 2: should we provide more details to make customers less worried? E.g. the loss of accuracy is barely noticeable.

cc @jan--f

Author: @simonpasquier and @jan--f  - I'd be fine with just omitting this sentence.

Reviewer 2: works for me.

[id="ocp-4-15-monitoring-improved-staleness-handling-for-the-kubelet-service-monitor"]
==== Improved staleness handling for the kubelet service monitor
With this release, staleness handling for the kubelet service monitor has been improved in order to ensure that alerts and time aggregations are accurate.
This improved functionality is active by default and makes the dedicated service monitors feature obsolete.

> As a result, the dedicated service monitors feature is now deprecated and will be removed in a future {product-title} release.

Reviewer 3: ```suggestion
As a result, the dedicated service monitors feature is now deprecated and will be removed in a future release. In this version, the setting has no effect.
```

Author: Edited to be more specific that the setting now has no effect in this release

[id="ocp-4-15-monitoring-improved-handling-of-user-workload-monitoring-component-issues"]
==== Improved handling of user workload monitoring component issues
This release improves how the system handles issues with pods or other components in user-defined projects.
These issues no longer cause the Cluster Monitoring Operator (CMO) to report that core platform monitoring components are in a degraded state.

> Previously, if a monitoring component in a user-defined project experienced problems, the CMO would report that it was in a degraded state even though the CMO was not in a degraded state. 

Reviewer: according to case [OCP-69084](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-69084), if a monitoring component in a user-defined project experienced problems, the CMO would still report that it was in a degraded state even though the CMO was not in a degraded state. @Tai-RedHat please help to confirm and update case if needed
```
$ oc get co monitoring 
NAME         VERSION                              AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
monitoring   4.15.0-0.nightly-2024-02-07-062935   False       True          True       6m57s   UpdatingUserWorkloadPrometheus: client rate limiter Wait returned an error: context deadline exceeded, UpdatingUserWorkloadThanosRuler: waiting for ThanosRuler object changes failed: waiting for Thanos Ruler openshift-user-workload-monitoring/user-workload: context deadline exceeded
```

Reviewer 3: @juzhao Thanks for reminding me. In fact, the description here needs to be modified. CC: @bburt-rh 

see from:
https://issues.redhat.com/browse/OCPBUGS-23748
``
This isn't what was implemented at the end (I changed the epic title + name to reflect that but not the description to keep the first idea but I will add a line, we talk about that in the comments though), what we did is make it easier and more reliable to detect degradations/unavailabilities due to UWM failures.
``

Author: @Tai-RedHat - I updated the description per my understanding of what was implemented. PTAL and lmk what you think. Thanks!

==== Updates to monitoring stack components and dependencies
This release includes the following version updates for in-cluster monitoring stack components and dependencies:

* Alertmanager to 0.26.0
* kube-state-metrics to 2.10.1
* node-exporter to 1.7.0
* Prometheus to 2.48.0
* Prometheus Adapter to 0.11.2
* Prometheus Operator to 0.70.0
* Thanos Querier to 0.32.5

[id="ocp-4-15-monitoring-changes-to-alerting-rules"]
==== Changes to alerting rules

[NOTE]
====
Red{nbsp}Hat does not guarantee backward compatibility for recording rules or alerting rules.
====

* The `NodeClockNotSynchronising` and `NodeClockSkewDetected` alerting rules are now disabled when the Precision Time Protocol (PTP) is in use.

[id="ocp-4-15-monitoring-new-metrics-server-to-access-metrics-API-technology-preview"]
==== New Metrics Server component to access the Metrics API (Technology Preview)
This release introduces a Technology Preview option to add a Metrics Server component to the in-cluster monitoring stack.
As a Technology Preview feature, Metrics Server is automatically installed instead of Prometheus Adapter if the `FeatureGate` custom resource is configured with the `TechPreviewNoUpgrade` option.

> If installed, Metrics Server, rather than Prometheus Adapter, collects resource metrics and exposes them in the `metrics.k8s.io` Metrics API service for use by other tools and APIs.

Reviewer 2: (nit) it reads as if Prometheus adapter collects metrics too which isn't the case.
```suggestion
If installed, Metrics Server collects resource metrics and exposes them in the `metrics.k8s.io` Metrics API service for use by other tools and APIs.
```

Author: Removed references to Prometheus Adapter

To use this feature, configure remote write using the `sendExemplars` option in the `RemoteWriteSpec` resource.
For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#remotewritespec[RemoteWriteSpec] in the config map API reference for the Cluster Monitoring Operator.

[id="ocp-4-15-monitoring-impoved-alert-querying-for-user-defined-projects"]
==== Improved alert querying for user-defined projects
Applications in user-defined projects now have API access to query alerts for application namespaces via the rules tenancy port for Thanos Querier. 
You can now construct queries that access the `/api/v1/alerts` endpoint via port 9093 for Thanos Querier, provided that the HTTP request contains a `namespace` parameter.
In previous releases, the rules tenancy port for Thanos Querier did not provide API access to the `/api/v1/alerts` endpoint.

[id="ocp-4-15-monitoring-prometheus-updated-to-tolerate-jitters-at-scrape-time"]
==== Prometheus updated to tolerate jitters at scrape time
The default Prometheus configuration in the monitoring stack has been updated so that jitters are tolerated at scrape time.

> For monitoring deployments that have showed sub-optimal chunk compression for data storage, this update helps to optimize data compression, thereby reducing the disk space used by the time series database in these deployments.

Reviewer: ```suggestion
For monitoring deployments that have shown sub-optimal chunk compression for data storage, this update helps to optimize data compression, thereby reducing the disk space used by the time series database in these deployments.
```

Author: 'shown` is the correct form with the particple: https://www.thesaurus.com/e/grammar/showed-or-shown/

[id="ocp-4-15-monitoring-improved-staleness-handling-for-the-kubelet-service-monitor"]
==== Improved staleness handling for the kubelet service monitor
Staleness handling for the kubelet service monitor has been improved in order to ensure that alerts and time aggregations are accurate.
This improved functionality is active by default and makes the dedicated service monitors feature obsolete.
As a result, the dedicated service monitors feature has been disabled and is now deprecated, and setting the `DedicatedServiceMonitors` resource to `enabled` has no effect.

[id="ocp-4-15-monitoring-improved-ability-to-troubleshoot-reports-of-tasks-failing"]
==== Improved ability to troubleshoot reports of tasks failing

> The reasons provided when tasks fail in monitoring components are now more granular so that you can more easily pinpoint whether a reported failure originated in components deployed in core platform monitoring, in user workload monitoring, or in both.

Reviewer 2: I haven't checked exhaustively but we generally use "user-defined monitoring" instead of "user workload monitoring" in the docs.

If the Cluster Monitoring Operator (CMO) reports task failures, you now see reasons that identify where the failures originated:

> * The `PlatformTasksFailed` reason indicates failures that originated with core platform monitoring components.

Reviewer: ```suggestion
* The `PlatformTasksFailed` reason indicates failures that originated with core platform monitoring components
```

Author: fixed

* The `UserWorkloadTasksFailed` reason indicates failures that originated with user workload monitoring components

> * The `MultipleTasksFailed` reason indicates failures that originated in both core platform monitoring and user workload monitoring components

Reviewer 2: (nit) there can be other reason values when a single component task fails (e.g. `UserWorkloadPrometheusFailed` or `PrometheusFailed`). When the reason is prefixed by `UpdatingUserWorkload`, it's about a UWM component otherwise it's platform. Not sure how you want to phrase it.
cc @machine424 to keep me honest.

Author: I re-worded it so that we don't imply that these are the only possible reason, i.e. that these reasons are just additional reasons added to help with troubleshooting.

For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#remotewritespec[RemoteWriteSpec] in the config map API reference for the Cluster Monitoring Operator.

[id="ocp-4-15-monitoring-impoved-alert-querying-for-user-defined-projects"]
==== Improved alert querying for user-defined projects
Applications in user-defined projects now have API access to query alerts for application namespaces via the rules tenancy port for Thanos Querier. 
Specifically, you can now construct queries that access the `/api/v1/alerts` endpoint via port 9093 for Thanos Querier, provided that the HTTP request contains a `namespace` parameter.
In previous releases, the rules tenancy port for Thanos Querier did not provide API access to the `/api/v1/alerts` endpoint.

[id="ocp-4-15-monitoring-prometheus-updated-to-tolerate-jitters-at-scrape-time"]
==== Prometheus updated to tolerate jitters at scrape time to reduce TSDB storage requirements
The default Prometheus configuration in the in-cluster monitoring stack has been changed so that jitters are tolerated at scrape time, which improves chunk compression for data storage.

> As a result, the time series database now uses significantly less disk space.

Author: @jan--f I dialed things back in the current version and added context.

If installed, Metrics Server collects resource metrics and exposes them in the `metrics.k8s.io` Metrics API service for use by other tools and APIs.
Using Metrics Server instead of Prometheus Adapter frees the core platform Prometheus stack from handling this functionality.
For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#metricsserverconfig[MetricsServerConfig] in the config map API reference for the Cluster Monitoring Operator.

[id="ocp-4-15-monitoring-new-feature-to-send-exemplar-data-to-remote-write-storage-for-user-defined-projects"]
==== New feature to send exemplar data to remote write storage for user-defined projects
User-defined projects can now use remote write to send exemplar data scraped by Prometheus to remote storage.
To use this feature, configure remote write using the `sendExemplars` option in the `RemoteWriteSpec` resource.
For more information, see xref:../monitoring/config-map-reference-for-the-cluster-monitoring-operator.adoc#remotewritespec[RemoteWriteSpec] in the config map API reference for the Cluster Monitoring Operator.

> [id="ocp-4-15-monitoring-impoved-alert-querying-for-user-defined-projects"]

Reviewer: ```suggestion
[id="ocp-4-15-monitoring-improved-alert-querying-for-user-defined-projects"]
```

Author: fixed

==== Improved staleness handling for the kubelet service monitor
Staleness handling for the kubelet service monitor has been improved in order to ensure that alerts and time aggregations are accurate.
This improved functionality is active by default and makes the dedicated service monitors feature obsolete.
As a result, the dedicated service monitors feature has been disabled and is now deprecated, and setting the `DedicatedServiceMonitors` resource to `enabled` has no effect.

[id="ocp-4-15-monitoring-improved-ability-to-troubleshoot-reports-of-tasks-failing"]
==== Improved ability to troubleshoot reports of tasks failing
The reasons provided when tasks fail in monitoring components are now more granular so that you can more easily pinpoint whether a reported failure originated in components deployed in the `openshift-monitoring` namespace or in the `openshift-user-workload-monitoring` namespace.
If the Cluster Monitoring Operator (CMO) reports task failures, the following reasons have been added to identify where the failures originated:

* The `PlatformTasksFailed` reason indicates failures that originated in the `openshift-monitoring` namespace.

> * The `UserWorkloadTasksFailed` reason indicates failures that originated in the `openshift-user-workload-monitoring` namespace.

Reviewer: in previous versions, there is also explanation for `MultipleTasksFailed`, but removed now, is that expected?
`The `MultipleTasksFailed` reason indicates failures that originated in both core platform monitoring and user-defined projects.`

Reviewer 4: yes, was removed because of: https://github.com/openshift/openshift-docs/pull/71003#discussion_r1484277019 and https://github.com/openshift/openshift-docs/pull/71003#discussion_r1484573918

This improved functionality is active by default and makes the dedicated service monitors feature obsolete.
As a result, the dedicated service monitors feature has been disabled and is now deprecated, and setting the `DedicatedServiceMonitors` resource to `enabled` has no effect.

[id="ocp-4-15-monitoring-improved-ability-to-troubleshoot-reports-of-tasks-failing"]
==== Improved ability to troubleshoot reports of tasks failing
The reasons provided when tasks fail in monitoring components are now more granular so that you can more easily pinpoint whether a reported failure originated in components deployed in core platform monitoring, in user-defined projects, or in both.
If the Cluster Monitoring Operator (CMO) reports task failures, the following reasons have been added to identify where the failures originated:

* The `PlatformTasksFailed` reason indicates failures that originated in core platform monitoring.
* The `UserWorkloadTasksFailed` reason indicates failures that originated in user-defined projects.
* The `MultipleTasksFailed` reason indicates failures that originated in both core platform monitoring and user-defined projects.

> 

Author: @machine424 - Can you PTAL at the wording in this section and help me make sure it's accurate and complete?

Reviewer 4: Talking about the new reasons only is a good idea.

How about:

```suggestion
The reasons provided when tasks fail in monitoring components are now more granular so that you can more easily pinpoint whether a reported failure originated in components deployed in core platform monitoring, or user-defined monitoring.
If the Cluster Monitoring Operator (CMO) reports task failures, the following reasons have been added to identify where the failures originated:

* The `PlatformTasksFailed` reason indicates failures that originated in core platform monitoring.
* The `UserWorkloadTasksFailed` reason indicates failures that originated in user-defined monitoring.
* The `MultipleTasksFailed` reason indicates failures that originated in both core platform monitoring and user-defined projects.
```

if you don't want to repeat `core platform monitoring`/`user-defined monitoring`, we can use the corresponding namespaces/projects: `openshift-monitoring`/`openshift-user-workload-monitoring`

Author: Updated the text per your suggestion.
